Remotion/Validation/Core/Validators/NotEmptyStringValidator.cs

```diff
@@ -0,0 +1,72 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using Remotion.FunctionalProgramming;
+using Remotion.Utilities;
+using Remotion.Validation.Implementation;
+using Remotion.Validation.Results;
+
+namespace Remotion.Validation.Validators
+{
+  public class NotEmptyStringValidator : IPropertyValidator
+  {
+    public string ErrorMessage { get; }
+    public ValidationMessage ValidationMessage { get; }
+
+    public NotEmptyStringValidator (ValidationMessage validationMessage)
+    {
+      ArgumentUtility.CheckNotNull("validationMessage", validationMessage);
+
+      ErrorMessage = "The value must not be empty.";
+      ValidationMessage = validationMessage;
+    }
+
+    public IEnumerable<PropertyValidationFailure> Validate (PropertyValidatorContext context)
+    {
+      if (IsValid(context))
+        return Enumerable.Empty<PropertyValidationFailure>();
+
+      return EnumerableUtility.Singleton(CreateValidationError(context));
+    }
+
+    private bool IsValid (PropertyValidatorContext context)
+    {
+      var propertyValue = context.PropertyValue;
+
+      if (propertyValue == null)
+        return true;
+
+      if(propertyValue is string stringValue)
+        return stringValue.Length > 0;
+
+      throw new NotSupportedException($"Only properties of type '{nameof(String)}' are supported.");
```

Comment:
Same as with NotEmptyCollectionValidator: incorrect type should be a valid result.

Reply:
Using a validator for an incompatible property is a programming error, isn't it? If so, the framework should support the developer in finding that he made a mistake.
Silently failing to validate anything and returning that the value is okay would hide the error. Why is that a better policy?

Reply:
Yes, that's the conventinal wisdom. Thing is, right now there's prior art with the other validator types, so we should keep the existing semantics. If it turns out that it's a bad idea / we want something else, we should make a separate task to change all the validators in one go.

In practice: it's a guardrail for the infrastructure itself and anyone who's building their own factory infrastructure. 

Historically: this descision was inherited from FluentValidation. The big question is, should a null-value and an invalid type be treated differently? It does feel more consistent to ignore all invalid values, otherwise you need to question why NULL is special.